### ISSUE TITLE
More flexible registers setup - unleash the power of SX127x radios

### ISSUE THREAD

**Issue by mostorer** [dernasherbrezon/sx127x#16]
Hello Andrey, I have downloaded you library from _ESPRegistry_ to integrate an SX127x radio with ESP32 devices. **Really a good work** but, it seems to me that it is particularly oriented towards Heltec and TTGo hardware devices. Of course, this is just my impression.

As an example, I have not found the way to have the radio working in Packet Mode (RegPacketConfig2 - bit 6 - DataMode), nor the way to granularly manage DIO mappings (e.g. I find useful to Data output to DIO5 and SyncAddress to DIO2 in test phases).

At this time I have made some quick and dirty changes in my code to get the result, but as a general purpose library on ESPRegistry it could be helpful to have some more control over SX127x functionalities.

Thank you in advance for you feedback.

**Comment by dernasherbrezon** [comment]
Hi,

Thank you for the feedback!

1. The library is very generic and tied to Heltec or TTGO (even though some examples based on them). See example for Raspberry PI & linux: https://github.com/dernasherbrezon/sx127x/blob/main/examples/receive_lora_raspberrypi/main.c
2. sx127x fully supports packet mode. Fixed length, variable length &etc. See some details here: https://github.com/dernasherbrezon/sx127x/blob/main/src/sx127x.c#L1269 
3. You can't output data on DIO5. According to the specification it is used to output clock.
```
5.3.2. CLKOUT Output
The reference frequency, or a fraction of it, can be provided on DIO5 (pin 13) by modifying bits ClkOut in RegDioMapping2.
```

I don't understand ```SyncAddress to DIO2```. Do you refer to "Continuous mode"? This mode is indeed is not supported.

**Comment by mostorer** [comment]
Hi, thank you fro your prompt response.

> The library is very generic and not tied to Heltec or TTGO (even though some examples based on them). See example for Raspberry PI & linux: https://github.com/dernasherbrezon/sx127x/blob/main/examples/receive_lora_raspberrypi/main.c

I will give a look; at this time only looked at `receive_fsk` and digged into your source code, but looking around it seems to me that some registers are not fully addressed nor customizable.

About the second point, not sure we are talking about the same thing. Packet Mode has to be enabled setting DataMode bit to 1 in RegPacketConfig2 (0x31). In your source code cannot find the way to do it.

> You can't output data on DIO5. According to the specification it is used to output clock.

5.3.2 paragraph is saying you can have `clkout on DIO5` ... but that is not the only DIO5 functionality ... please, give a look at "`Table 30 DIO Mapping, Packet Mode`"

Last but not least, `SyncAddress to DIO2` is defined in the same table 30.

**Comment by dernasherbrezon** [comment]
> some registers are not fully addressed nor customizable.

Yes, library is adding very thin abstraction. But it is still abstraction. If you want full control, then better manipulate them directly without library. Or if you have a specific use-case, I can add into the library.

> About the second point, not sure we are talking about the same thing.

Yes, we're on the same page. DataMode is 0x01 (packet mode) by default. And sx127x don't change it. I can set it to 0x01 ( https://github.com/dernasherbrezon/sx127x/blob/main/src/sx127x.c#L1269) to explicitly indicate packet mode.

> that is not the only DIO5 functionality

Can you describe your use-case? If you're bypassing packet assembly by sx127x, then maybe you don't need a library?

**Comment by mostorer** [comment]
> Can you describe your use-case? If you're bypassing packet assembly by sx127x, then maybe you don't need a library?

Not bypassing the packet assembly otherwise, you are right, why to use the library ?
As I wrote, working with variable packets from different sources (not always well-known), during development and debugging phase I find useful to have the received data (and other signals) on DIOs, available for a logic analyzer.

As an example, in the picture below you can see Data, Preamble Detect, Sync Detect and CRC ok signals.

![image](https://github.com/user-attachments/assets/73b45585-30b8-472d-983b-29da4f9120b7)

This way I can be sure the configuration and the code are correct.

**Comment by mostorer** [comment]
> This is an io-homecontrol frame. This library is working well without modification for that

It seems also this time I cannot agree with you. Anyway, the topic is not related to that, but to the possibility to have more low level control over sx127x registries.

**Comment by dernasherbrezon** [comment]
I'm thinking about several options:

1. Do nothing. You can actually read/write registers using low_level functions from sx127x_spi.h
2. Add support for continuous mode. Reading/Writing data directly from the pin is used in the continuous mode. I guess it won't hurt to additionally configure sync address, preamble detect and crc interrupts.
3. Add support for bulk writes/reads. I don't have such use-cases, but it might work like this:
   * Configure registers individually using functions provided
   * Test setup in the lab
   * Once everything tested, read snapshot of all registers into uint8_t[128].
   * When running in production simply write that array in one go. It would cause less SPI transfers and be faster.

**Comment by dernasherbrezon** [comment]
Hmm.

> they should be moved to include file

This will affect public API. Your use case about troubleshooting things before going into production. Why can't use simply do:

```c
sx127x_spi_write_register(0x31, 0b11110000, 1, spi_device);
```

If it is throwaway and temporary code for some research, then it can use hardcoded values.

> The same would be done for registers values and masks

If you want low level, then you already know these registers and masks. Why expose them?

> ```shadow_spi_device_t``` should be also moved to include file

Actually I already did it in the upcoming 4.0 version:
https://github.com/dernasherbrezon/sx127x/blob/c94ba726afa63220ca67059c059216cd55b1d3ec/include/sx127x.h#L326

But for different purpose. New version will allow initialise memory statically and have tuned memory with special pre-processor directives.

**Comment by mostorer** [comment]
> If you want low level, then you already know these registers and masks. Why expose them?

I think all depends on how you intend your library to be used ...

1. If your goal is to block their use to only the high level APIs currently exposed, then there is nothing more to do; this way, if someone wants to use them to make something not decided/supported by you, they are useless.

2. if your goal is instead to simplify the use of sx127x devices through high level APIs, but leaving the ability to more skilled people to use them also in more flexible way for scenarios not currently thought by you, then also low level interface has to be exposed (including registries definitions).

Up to you.

**Comment by dernasherbrezon** [comment]
Do you have examples of similar libraries with such low level access?

**Comment by mostorer** [comment]
I think it's not a question of what other libraries do, but what you want your libraries to do.

**Comment by Velocet** [comment]
I understand your points and maybe a compromise is the way to go:
Download the SX1276 app from Semtech.

- Press CTRL+ALT+N to unlock.
- Export the registers.
- Configure the radio.
- Export the registers again.
- Compare both exports and use the low level functions to alter the registers.

**Comment by mostorer** [comment]
Getting the Semtech registers header file is not so difficult so, if you add `shadow_spi_device_t` in your header file, it will be easy to use low level functions adding Semtech header as an additional include.

Are you going to add `sx127x_shadow_spi_write_register`, `sx127x_read_register`and `sx127x_read_register` in your header file or they will have to be re-defined externally ?

**Comment by dernasherbrezon** [comment]
Why do you need `shadow_spi_device_t`? You can use existing code to read/write registers directly.

```c
#include "sx127x_spi.h"

spi_device_handle_t spi_device;
spi_bus_add_device(SPI2_HOST, &dev_cfg, &spi_device);
sx127x_create(spi_device, &device);
...

uint32_t result;
sx127x_spi_read_registers(0x31, spi_device, 1, &result);
sx127x_spi_write_register(0x31, &result, 1, spi_device);
```

**Comment by mostorer** [comment]
I understood, but maybe I am wrong, you are managing a cache of some registers through `sx127x_shadow_spi_write_register`. 
If so, using `sx127x_spi_write_register ` could lead to an out-of-sync cache. Isn't it ?

**Comment by dernasherbrezon** [comment]
Added 2 things:

- Separate include file with all registers b8b38884be81edb03c77cfabb12809e975b07b9c
- `sx127x_read_register` and `sx127x_write_register` for direct register manipulation 3f575c28a3fe19f54276ec6d532a93defc6d7ee6

**Comment by mostorer** [comment]
Well done!
Are you going to release this changes also to Espressif Component Registry ?

**Comment by dernasherbrezon** [comment]
I just released new version - 4.0.0. It should include this changes. Please note it is not backward compatible with 3.x. Migration is trivial, but still needed.

**Comment by Velocet** [comment]
Just a quick note about custom DIO mappings:

Do NOT enable clock output if you are not using it as it can cause RF interference.
